integration/setup.go

```diff
@@ -0,0 +1,161 @@
+	// for our tests; these relative paths break when we run the binary tests. To get around this,
+	// at time of compilation, we save the path to the source code so that we can reference it
+	// when running the tests. This is hacky as fuck but works!
+	// Alternative was to hard code the configs in code but that's not possible as most of the configs
+	// are private.
+	basePath string
+)
+
+type testServerSetup struct {
+	db   storage.Database
+	opts instrument.Options
+	cfg  configuration
+
+	// Signals.
+	doneCh   chan struct{}
+	closedCh chan struct{}
+}
+
+func newTestServerSetup(t *testing.T, configFile string) *testServerSetup {
+	configFile = basePath + configFile
+	var cfg configuration
+	if err := xconfig.LoadFile(&cfg, configFile, xconfig.Options{}); err != nil {
+		t.Fatalf("error loading config file %s: %v\n", configFile, err)
+	}
+
+	iOpts := instrument.NewOptions().
+		SetMetricsScope(tally.NoopScope).
+		SetLogger(logger)
+
+	namespaces, err := cfg.Database.NewNamespacesMetadata()
+	if err != nil {
+		t.Fatalf("error creating namespaces metadata: %v", err)
+	}
+	shardSet, err := cfg.Database.NewShardSet()
+	if err != nil {
+		t.Fatalf("error creating shard set: %v", err)
+	}
+	dbOpts, err := cfg.Database.NewOptions(iOpts)
+	if err != nil {
+		t.Fatalf("error creating database options: %v", err)
+	}
+	db := storage.NewDatabase(namespaces, shardSet, dbOpts)
+	if err := db.Open(); err != nil {
+		t.Fatalf("error opening database: %v", err)
+	}
+	return &testServerSetup{
+		db:       db,
+		opts:     iOpts,
+		cfg:      cfg,
+		doneCh:   make(chan struct{}),
+		closedCh: make(chan struct{}),
+	}
+}
+
+func (ts *testServerSetup) startServer() {
+	go func() {
+		// TODO (wjang): pass in 0.0.0.0:0 instead, have an automatically generated port and use it.
+		if err := serve.Serve(
+			ts.cfg.HTTP.ListenAddress,
+			ts.cfg.HTTP.Handler.NewOptions(ts.opts),
+			ts.cfg.HTTP.NewServerOptions(),
+			ts.db,
+			logger,
+			ts.doneCh,
+		); err != nil {
+			logger.Fatalf("could not start serving traffic: %v", err)
+		}
+		close(ts.closedCh)
+	}()
+}
+
+func (ts *testServerSetup) stopServer(t *testing.T) {
+	close(ts.doneCh)
+
+	select {
+	case <-ts.closedCh:
+		t.Log("server closed clean")
+	case <-time.After(gracefulShutdownTimeout):
+		t.Logf("server closed due to %s timeout", gracefulShutdownTimeout.String())
+	}
+}
+
+func (ts *testServerSetup) stopDB(t *testing.T) {
+	// TODO(wjang): delete the database files as well
+	// TODO(wjang): Close() is panic-ing in storage/shard.go
+	err := ts.db.Close()
+	if err != nil {
+		t.Fatalf("error closing database: %v", err)
+	}
+}
+
+func (ts *testServerSetup) close(t *testing.T) {
+	ts.stopServer(t)
+	ts.stopDB(t)
+}
+
+func (ts *testServerSetup) newClient() client {
+	client := newClient(ts.cfg.HTTP.ListenAddress)
+	for {
+		if client.serverIsHealthy() {
+			break
+		}
+		time.Sleep(time.Millisecond * 10)
+	}
+	return client
+}
+
+func (ts *testServerSetup) writeTestFixture(t *testing.T) {
```

Comment:
Hm this reads the input config and writes it to fixture? Probably just store the config in a var as suggested above as this seems really hacky.

Reply:
Remove this perhaps? There is no need to stash the test data inside the test setup. You can just call `client.write([]byte(testData))` inside the test so this is redundant.